### PR TITLE
Refactor and simplify PWADevServer host/port handling

### PR DESCRIPTION
**Pull request by zetlen**
## This PR is a:

[ ] New feature
[x] Enhancement/Optimization
[x] Refactor
[ ] Bugfix
[ ] Test for existing code
[ ] Documentation

## Summary

When this pull request is merged, it will pull out a lot of our hand-rolled hostname and SSL certificate handling, in favor of davewasmer/devcert which has made great strides and handles most of what we want to handle.

- Refactored the host generation into a `configureHost` utility exposed in Buildpack. Unique host generation can now be shared by both dev and staging.
- Changed the API for secure hosts.
  - Now, there are not separate options for unique unsecure hosts and secure certificates. If you opt in to the provided host, you'll get SSL.
  - Instead of being randomly generated and then stored in a database, the hostnames and ports are derived from a hash of the current directory; no persistence is required.
- **There is no more stateful database**. The `~/.config/pwa-buildpack.db` is gone for now. No promises on whether it will return.
- The staging server now uses the same `configureHost` utility to create a unique url.
- Removed many utils and tests which were no longer necessary.
- To share configuration, the webpack.config.js and staging server now obtain their `provideSecureHost` config from documented environment variables, which have been added to `example.env`.

## Additional information

Some rules of thumb about the default implementation:  

- We recommend simply setting `MAGENTO_BUILDPACK_PROVIDE_SECURE_HOST=1` and `MAGENTO_BUILDPACK_SECURE_HOST_ADD_UNIQUE_HASH=1`. This produces readable domains and unique ports that work for most cases.
- Generally, a port in the 8000-8999 range indicates a dev server, a port in the 9000-9999 range indicates a staging server, and a port 10000 or above indicates a temporary port because the preferred one is in use.
- The staging server will not automatically select an in-use port. It is meant to be configured externally, by provisioning scripts. 
- In order to get *anything* to work, I had to discard the `lerna link convert` methodology that had caused so many conflicts. Many of the changes are just reorganizing dependency config with no functionality change.

### CODE REVIEW

**Comment by zetlen** [comment]
The `package-lock.json` files are back, so this looks like a huge PR, but it isn't one. Lots of homegrown code was removed.

**Comment by zetlen** [comment]
@mhhansen If you want to help today, I would be real grateful if you could take a look at this one. This one is an enabler of other people getting work done, as I think it removes a lot of bugs and a lot of code burden from us.

**Comment by mhhansen** [comment]
**1. works and servers OK**
OK `watch:venia`, creates, runs and serves without troubles or errors related to SSL.
I was asked for password to store the local domain in _hosts file_, but just the first time.
which ended up serving OK:
```
PWADevServer ready at https://magento-venia-concept-cryj2.local.pwadev:8533
```


**2. You first needs to run interactive mode**
You can't run `watch:all` if current domain was not previously added to the hosts file
It will ask for `sudo` pass to store the domain. so, needs to be interactive, otherwise it _complains_ about that, and suggest to first run the _dev environment_ (`watch:venia`). Once it's added
This will happen, each time the domain changes (cause it needs to be added into the `hosts` file)
Makes sense, but this behavior I think should be added to devdocs.


**3. (improve) declare a default subdomain to avoid `undefined` when unique_hash is not present**
While having only`secure_host` enabled (without `UNIQUE_HASH`), it will create an `undefined` subdomain.
```
MAGENTO_BUILDPACK_PROVIDE_SECURE_HOST=1
```
screen shot 2018-10-09 at 8 31 20 am

If you provide the `secure_host_subdomain` option (and comment out the `unique_hash`), it creates and make use of the specified subdomain:
```
MAGENTO_BUILDPACK_PROVIDE_SECURE_HOST=1
MAGENTO_BUILDPACK_SECURE_HOST_SUBDOMAIN="bjork"
#MAGENTO_BUILDPACK_SECURE_HOST_ADD_UNIQUE_HASH=1
```
screen shot 2018-10-09 at 8 35 35 am



**4. add to venia .env file**
We've to add and document into the `pacakges/venia-concept/example.env` new env options:
```
MAGENTO_BUILDPACK_PROVIDE_SECURE_HOST
MAGENTO_BUILDPACK_SECURE_HOST_ADD_UNIQUE_HASH
MAGENTO_BUILDPACK_SECURE_HOST_SUBDOMAIN
MAGENTO_BUILDPACK_SECURE_HOST_EXACT_DOMAIN
```


**5. `vendor.js` stills gets called**
```
bjork.local.pwadev/:14 GET https://bjork.local.pwadev:8533/js/vendor.js net::ERR_ABORTED 500 (Internal Server Error)
```

Couldn't reproduce any of the `ERR_SSL_PROTOCOL_ERROR` and so..
I'll add/send code review for points 3 and 4, just wanted to deliver the feedback.

I'll run a few more tests with SW enabled to see how it behaves. Cause I ended up with MagentoRouteHandler error, from the `checkPropTypes` expecting number got string, error. but it only happened with my instance that has `venia sample data` installed. tried out with a second m2 install that has the `default sample data` (installed via CLI `sampledata:deploy`), and I don't get that error.


**BTW: tested over Win10**, and after tweaking a few scripts from `package.json` I was able to `npm install`. the thing is, there's one package that executes unix commands, so you can't run the servers currently. will take a look later this week.

**Comment by zetlen** [comment]
That certainly was a fun CI adventure.

File: packages/venia-concept/webpack.config.js [hunk 1]
```diff
@@ -133,14 +133,20 @@ module.exports = async function(env) {
...
-            id: 'magento-venia',
-            provideSSLCert: true
-        });
+        const devServerConfig = {
+            publicPath: config.output.publicPath
+        };
+        const provideHost = !!process.env.MAGENTO_BUILDPACK_PROVIDE_SECURE_HOST;
+        if (provideHost) {
+            devServerConfig.provideSecureHost = {
+                subdomain: process.env.MAGENTO_BUILDPACK_SECURE_HOST_SUBDOMAIN,
```
[the remark this answers is not in the document]
Comment (mhhansen) [reply packages/venia-concept/webpack.config.js]: just checked, yes you did 👍

**Pull request merged**